### DISCUSSION TITLE
Creating a custom board definition growing pains, what did i miss?

### DISCUSSION THREAD

**Discussion by Jibun-no-Kage** [micropython/micropython#9207] [Core Development] [answered]
Creating a custom board definition growing pains, what did i miss?

I have a VM that has the SDK (ESP) and clone MP sources, I can generated any port I need.  So now time to customize.

I copied the PICO_W board definition to PICO_W_DD, customized the manifest.  I then tried to generate the firmware image and hit the following error.  Clearly I missed something somewhere.

```
crosstoolng@psycho:~/micropython/ports/rp2$ make clean
rm -f -rf build-PICO
crosstoolng@psycho:~/micropython/ports/rp2$ make -j 2 BOARD=PICO_W_DD
[ -e build-PICO_W_DD/Makefile ] || cmake -S . -B build-PICO_W_DD -DPICO_BUILD_DOCS=0 -DMICROPY_BOARD=PICO_W_DD
PICO_SDK_PATH is /home/crosstoolng/micropython/lib/pico-sdk
PICO platform is rp2040.
Build type is MinSizeRel
PICO target board is pico_w_dd.
CMake Error at /home/crosstoolng/micropython/lib/pico-sdk/src/boards/generic_board.cmake:22 (message):
  Unable to find definition of board 'pico_w_dd' (specified by PICO_BOARD):

     Looked for pico_w_dd.h in /home/crosstoolng/micropython/lib/pico-sdk/src/boards/include/boards (additional paths specified by PICO_BOARD_HEADER_DIRS)
     Looked for pico_w_dd.cmake in /home/crosstoolng/micropython/lib/pico-sdk/src/boards (additional paths specified by PICO_BOARD_CMAKE_DIRS)
Call Stack (most recent call first):
  /home/crosstoolng/micropython/lib/pico-sdk/src/board_setup.cmake:28 (include)
  /home/crosstoolng/micropython/lib/pico-sdk/src/CMakeLists.txt:15 (include)


-- Configuring incomplete, errors occurred!
See also "/home/crosstoolng/micropython/ports/rp2/build-PICO_W_DD/CMakeFiles/CMakeOutput.log".
See also "/home/crosstoolng/micropython/ports/rp2/build-PICO_W_DD/CMakeFiles/CMakeError.log".
make: *** [Makefile:26: all] Error 1
```

Custom board directory...

```
crosstoolng@psycho:~/micropython/ports/rp2/boards/PICO_W_DD$ ls -l
total 16
-rw-r--r-- 1 crosstoolng crosstoolng 424 Sep  1 23:34 board.json
-rw-r--r-- 1 crosstoolng crosstoolng 329 Sep  2 17:51 manifest.py
-rw-r--r-- 1 crosstoolng crosstoolng 254 Sep  2 18:24 mpconfigboard.cmake
-rw-r--r-- 1 crosstoolng crosstoolng 903 Sep  2 17:55 mpconfigboard.h
crosstoolng@psycho:~/micropython/ports/rp2/boards/PICO_W_DD$
```

The manifest.py...

```
#include("../manifest.py")
include("$(MPY_DIR)/extmod/uasyncio/manifest.py")

#freeze("$(MPY_DIR)/tools", "upip.py")
#freeze("$(MPY_DIR)/tools", "upip_utarfile.py")
freeze("$(MPY_DIR)/extmod", "ntptime.py")

#if os.path.isdir(convert_path("$(MPY_LIB_DIR)")):
    #freeze("$(MPY_LIB_DIR)/python-ecosys/urequests", "urequests.py")
```

The mpconfigboard.h file...

```
// Board and hardware specific configuration
#define MICROPY_HW_BOARD_NAME                   "Raspberry Pi Pico W DD"

// todo: We need something to check our binary size
#define MICROPY_HW_FLASH_STORAGE_BYTES          (848 * 1024)

// Enable networking.
#define MICROPY_PY_NETWORK 1

// CYW43 driver configuration.
#define CYW43_USE_SPI (1)
#define CYW43_LWIP (1)
#define CYW43_GPIO (1)
#define CYW43_SPI_PIO (1)

// For debugging mbedtls - also set
// Debug level (0-4) 1=warning, 2=info, 3=debug, 4=verbose
// #define MODUSSL_MBEDTLS_DEBUG_LEVEL 1

#define MICROPY_HW_PIN_CYW43_COUNT CYW43_WL_GPIO_COUNT
#ifdef CYW43_WL_GPIO_LED_PIN
#define MICROPY_HW_PIN_CYW43_LED_PIN_NUM CYW43_WL_GPIO_LED_PIN
#endif

#define MICROPY_HW_PIN_RESERVED(i) ((i) == CYW43_PIN_WL_HOST_WAKE || (i) == CYW43_PIN_WL_REG_ON)

#define MICROPY_PY_USSL                 (0)
#define MICROPY_PY_UWEBSOCKET           (0)
#define MICROPY_PY_WEBREPL              (0)
```

The cmake file...

```
# cmake file for Raspberry Pi Pico W DD
set(MICROPY_BOARD PICO_W_DD)

set(MICROPY_PY_LWIP ON)
set(MICROPY_PY_NETWORK_CYW43 ON)

# Board specific version of the frozen manifest
set(MICROPY_FROZEN_MANIFEST ${CMAKE_SOURCE_DIR}/boards/PICO_W_DD/manifest.py)
```

**Comment by Jibun-no-Kage** [comment]
The fix seems straight forward, I just need to clone the header file as well in the Pico SDK.  But is there not a way to let the Pico SDK now that the base board is still a Pico W, but use the custom board definition layer on top of the base Pico W SDK header?

So... specific to the error above, since the Pico SDK is embedded in the MP source tree...

```
\# cd ~/micropython/lib/pico-sdk/src/boards/include/boards
\# cp pico_w.h pico_w_dd.h
```

How the make for the Pico W DD firmware image completes as expected.

**Comment by Lobo-T** [comment, marked as answer]
I think you can set:
`set(PICO_BOARD pico_w)`
In mpconfigboard.h.

**Reply by ricksorensen** [reply]
.... mpconfigboard.cmake I believe is the correct spot for this.

**Comment by jimmo** [comment]
@Jibun-no-Kage as @Lobo-T and @ricksorensen have pointed out, the issue here is that you have

```
set(MICROPY_BOARD PICO_W_DD)
```

in `mpconfigboard.mk`.

The meaning of this variable is that it's the board that the pico-sdk is using (they have a parallel concept of board definitions, but they are not 1:1). Hence why you had to add it to pico-sdk too. In your case, if your board is functionally identical to the official Pico W (as far as the pico-sdk is concerned) then you should just set it to `PICO_W`.

**Comment by mattytrentini** [comment]
Take a look at #8505 for a _strong hint_. Pico-sdk normally expects a board definition within the (pico-sdk) repo (see [boards](https://github.com/raspberrypi/pico-sdk/tree/master/src/boards/include/boards)) but they _can_ be defined out-of-repository. The key variable is `PICO_BOARD_HEADER_DIRS`, see [mpconfigboard.cmake:6](https://github.com/micropython/micropython/pull/8505/files#diff-d80c2398e084524321677ed0f9e23a67c4b02dbba0ec2619be081e05a3dc6ff5R6).

Note that #8505 is still under active development and is likely to be updated soon...

**Comment by Jibun-no-Kage** [comment]
I want to say I tried that, but, I could have not stage it correctly, in the process of testing and learning how this works, so will try it again.  Creating a new header file in the SDK would make sense for a 'new' board of course.  But really I want to use the existing SDK definition, but custom MP board definition.  Back to testing, thanks guys.

**Reply by Jibun-no-Kage** [reply]
Ok, got every right now, so the board chain is no longer broken, and SDK is using the existing Pico SDK, Pico W, header correctly, i.e. no custom header file needed.  I did block the default manifest file include to avoid the components I did not want to be included.  But I have not figured out, using this new approach how to disable the default display and frame buffer support.  My project is such that I will never have a display of any type connected to the Pico, so would like to remove that support.  But I don't see a 'clean' way to do that given the current architecture of the design.  In fact, there really is not as yet a way to control the extmod inclusions that is not a bit of kludge, right?

**Reply by jimmo** [reply]
> default display and frame buffer support

I'm not sure what the "default display support" is. But for framebuffer you need

```c
#define MICROPY_PY_FRAMEBUF (0)
```

in your board's `mpconfigboard.h`

**Reply by Jibun-no-Kage** [reply]
That works but other directives don't.  Interesting.  Some directives can't be turned off in rp2/mpconfigport.h file?  Even if defined in the board file, get compile/link errors?

#define MICROPY_PY_USSL                (0)
#define MICROPY_PY_UWEBSOCKET  (0)
#define MICROPY_PY_WEBREPL         (0)

And another question, under the 'new' method/documentation, will you be able to turn off features in the board definition, that actually may be defined a the port level or scope?

**Reply by jimmo** [reply]
You probably need to do a `make BOARD=... clean` first.

It looks like the `cmake` build used by the rp2 port doesn't have an equivalent of `QSTR_GLOBAL_DEPENDENCIES` which the other ports use to inject mpconfigboard.h as a dependency.

**Reply by Jibun-no-Kage** [reply]
Yup, and I am also pulling the clone down as well, since I have done so many tests, I may have something wacked elsewhere.  I wrote a few bash scripts to completely reset/stage everything, so easy to do this.  If I do have to keep board vs. port headers manually I can, not a big deal.
